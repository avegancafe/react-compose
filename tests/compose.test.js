import React, { Component, Fragment } from 'react'
import { mount } from 'enzyme'

import compose from '../index.js'

let wrapper

describe('#compose', () => {
  describe('with an empty array', () => {
    it('errors out', (done) => {
      const tmp = console.error;
      console.error = () => null

      try {
        mounted = mount(<AppWithoutRenderers />)
      } catch (e) {
        done()
      } finally {
        console.error = tmp
      }
    })
  })

  describe('with a list of component instances', () => {
    beforeEach(() => {
      wrapper = mount(<AppWithRenderers />)
    })

    it('passes args through to render function in order', () => {
      const expectedArgs = [
        props.mousePosition,
        props.subscription,
        props.validate
      ]

      expect(composedChildMock).toHaveBeenCalledWith(expectedArgs)
      expect(wrapper).toMatchSnapshot()
    })
  })
})

/* Fixtures */

const props = {
  mousePosition: { x: 300, y: 427 },
  subscription: { data: ["note 1", "note 2"] },
  validate: { predicate: x => (x < 10) }
}

const MousePosition = ({ children }) => (
  <Fragment>{children(props.mousePosition)}</Fragment>
)

const Subscription = ({ children }) => (
  <Fragment>{children(props.subscription)}</Fragment>
)

const Validate = ({ children }) => (
  <Fragment>{children(props.validate)}</Fragment>
)

const ComposedComponentWithRenderers = compose([
  <MousePosition relative={true} />,
  <Subscription url="/api/notes" />,
  <Validate />
])

const ComposedComponentWithoutRenderers = compose([])

const composedChildMock = jest.fn(() => <div>hi</div>)

const AppWithRenderers = () => (
  <ComposedComponentWithRenderers>
    {composedChildMock}
  </ComposedComponentWithRenderers>
)

const AppWithoutRenderers = () => (
  <ComposedComponentWithoutRenderers>
    {composedChildMock}
  </ComposedComponentWithoutRenderers>
)

