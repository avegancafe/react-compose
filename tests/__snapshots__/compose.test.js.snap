// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#compose with a list of component instances passes args through to render function in order 1`] = `
<AppWithRenderers>
  <Component>
    <ComposedWrapper>
      <MousePosition
        relative={true}
      >
        <ComposedWrapper
          __composedArgs={
            Array [
              Object {
                "x": 300,
                "y": 427,
              },
            ]
          }
        >
          <Subscription
            url="/api/notes"
          >
            <ComposedWrapper
              __composedArgs={
                Array [
                  Object {
                    "data": Array [
                      "note 1",
                      "note 2",
                    ],
                  },
                  Object {
                    "x": 300,
                    "y": 427,
                  },
                ]
              }
            >
              <Validate>
                <initial
                  __composedArgs={
                    Array [
                      Object {
                        "x": 300,
                        "y": 427,
                      },
                      Object {
                        "data": Array [
                          "note 1",
                          "note 2",
                        ],
                      },
                      Object {
                        "predicate": [Function],
                      },
                    ]
                  }
                >
                  <div>
                    hi
                  </div>
                </initial>
              </Validate>
            </ComposedWrapper>
          </Subscription>
        </ComposedWrapper>
      </MousePosition>
    </ComposedWrapper>
  </Component>
</AppWithRenderers>
`;
